Text displayed on a multiline braille display will now word wrap on every row in the braille window, not just at the end

michaelDCurran opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
None.

### Summary of the issue:
Although displaying a long line of braille on a multiline braille display fills up as much of the display as possible, rows may end in the middle of a word. 
NVDA should honor the 'avoid splitting words when possible' Braille setting for all rows on the display.
 
### Description of user facing changes
When displaying braille on a multiline braille display, NVDA will avoid splitting words over lines if the 'avoid splitting words when possible' braille setting is on.

### Description of development approach
Rather than tracking a braille window as only `_windowStartPos` and `_windowEndPos`, internally store a list of window rows which map to start and end offsets in the braille buffer. These offsets are calculated each time the window is moved or the buffer is updated, and ensure that words are not split across window rows.

This PR also adds / changes the filters available for external code such as Remote to dictate the size of the display:
* braille: Added a filter_displayDimensions extension point which takes a namedtuple of numRows and numCols, allowing external code to dictate the number of rows and columns for the display. This should be used now in place of filter_displaySize. This stops the worry of external code changing the number of rows but forgetting to change the over all size.
* braille.handler: Added a displayDimensions property which:
    * creates a DisplayDimensions namedtuple of the display's real numRows and numCols.
    * Filters this with filter_displayDimensions.
    * Calculates displaySize from the product of the filtered numRows and numCols.
    * If at least one handler is registered on the old displaySize filter:
        * Filters this with the old filter_displaySize.
        * If the displaySize changes due to filtering, for backwards compatibility the DisplayDimensions numRows is set to 1 and numCols is set to the filtered displaySize.
        * If the DisplayDimensions have changed at all from the internal cache after all the filters and calculations, then fire the handle_displaySizechanged extension point passing displaySize, numRows and numCols.

This all means that new NVDA / remote code could specify full display dimensions, but for backwards compatibility, if the old displaySize filter is used, then braille is forced to a single row with displaySize cells.

### Testing strategy:
* Tested NVDA with a Dot pad braille display. Used NVDA in various standard situations for a length of time, displaying long lines of documents and controls, ensuring that splitting words is avoided.
* Performed the same above test but by dictating display should have 3 rows and 10 columns via the displayDimensions filter. Ensured that content only filled a 3 by 10 space on the DotPad and nothing was cut off and I could still pan to see more.
* Performed the same above test but by dictating display should have a display size of 15 via the old displaySize filter. Ensured that content only filled a single line of 15 cells on the DotPad and nothing was cut off and I could still pan to see more.
* Tested with an Orbit Reader single line display, ensuring that displaying and cursor routing still function as before this PR.
 
### Known issues with pull request:
Code only applies the displaySize filter if at least one handler has registeref for the filter. this is checked by `len(list(filter_displaySize.handlers))`. There is a technical race condition here as a handler could unregister before the filter is applied. However the worse case is really just that a multiline display will stay singleline for the current display update.
In future it might be nice if the idea of checking for extension point Filter registrations could be standardized better, in particular some kind of atomic call  that both communicated if handlers were registered and filtered a value if so. 

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Improved handling of braille display windows with enhanced buffer management and row-based display logic.
	- Introduced a new method to convert window-relative positions to buffer-relative positions, offering better content control.

- **Bug Fixes**
	- Restructured display dimensions management for more accurate rendering, avoiding word splits in multi-row configurations.

- **Tests**
	- Added unit tests for resizing cell arrays to ensure functionality across various scenarios.
	- Updated existing tests to align with the new display dimensions structure.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

michaelDCurran committed:
BrailleBuffer: abstract conversion of a window pos to a buffer pos into its own windowPosToBufferPos method.

michaelDCurran committed:
BrailleBuffer: for displays with multiple rows of cells, word wrap each row, if word wrapping is enabled.

michaelDCurran committed:
Merge branch 'master' into brailleWordWrap

michaelDCurran committed:
Merge branch 'master' into brailleWordWrap

michaelDCurran committed:
Merge branch 'master' into brailleWordWrap

michaelDCurran committed:
Merge branch 'master' into brailleWordWrap

michaelDCurran committed:
Ensure _windowRowOffsets is initalized at least for one empty row at the start.

michaelDCurran committed:
linting

michaelDCurran commented:
I'm not too sure what to do about the displaySize property and the failing unit tests yet. Does displaySize as it is now, still make sense now we have numRows and numCols? I assume that displaySize is used for nvda remote for making the size of the two braille displays match. If so, what should we do in the case of multiline displays?

LeonarddeR reviewed (commented):
I think the problem with the unit tests lies in that the new code relies on numCols, wherea's the unit tests enable the braille handler with a registration to `braille.filter_displaySize`. In that case, numCols on the display will be 0.
I now also understand your question about the displaySize property. May be the handler should also have displayRows and displayCols auto properties.
- displayRows should just return `self.display.numRows`
- displayCols should return `self.displaySize // self.numRows`
In that case, backwards compatibility is preserved. However, when I control a remote machine with a 40 cell display attached with my machine that has a multi row display (e.g. two rows of 40 cells), we'll end up having 20 cells per row rather than only one row with 40 cells. This can be fixed by making the `displayRows` property a bit smarter:
- If `display.numRows` is 1 or displaySize > numCells, just return `numRows`
- If `display.numRows` > 1 and `displaySize` < `display.numCells`, return something like `min(self.display.numRows, math.ceil(self.displaySize / self.display.numCols))`
    This means that in the example above, a displaySize of 20 will use the top row, 40 will use the top row, 50 will use 40 cells of the top row and 10 cells of the second row.

bramd reviewed (commented):
In general this looks good to me. I'll test it on real hardware over the weekend.

bramd commented on source/braille.py:
```diff
@@ -1760,9 +1760,9 @@ def __init__(self, handler):
 		#: @type: [int, ...]
 		self.brailleCells = []
+		self._windowRowBufferOffsets: list[tuple[int, int]] = [(0, 0)]
 		#: The position in L{brailleCells} where the display window starts (inclusive).
 		#: @type: int
```
```suggestion
```

Remove left-over docstring for removed windowStartPos

michaelDCurran committed:
Merge branch 'master' into brailleWordWrap

michaelDCurran committed:
braille.py: add variable comment. Remove no longer needed comment.

michaelDCurran commented on source/braille.py:
```diff
@@ -1760,9 +1760,9 @@ def __init__(self, handler):
 		#: The translated braille representation of the entire buffer.
 		#: @type: [int, ...]
 		self.brailleCells = []
+		self._windowRowBufferOffsets: list[tuple[int, int]] = [(0, 0)]
```
Added above.

michaelDCurran committed:
Apply suggestions from code review

Co-authored-by: Bram Duvigneau <bram@bramd.nl>
Co-authored-by: Leonard de Ruijter <3049216+LeonarddeR@users.noreply.github.com>

michaelDCurran committed:
Merge branch 'master' into brailleWordWrap

michaelDCurran committed:
braille.BrailleBuffer.update: do not pad brailleCells up to display  length here. This breaks several unit tests and was not needed in the end.
 BrailleBuffer.windowStartPos setter: when calculating window row buffer offsets, stop once the end of the braille buffer is reached. This ensures that windowEndPos will always end at the end of the buffer, and not be padded out to the display length. This again fixes issues in several unit tests that assume that a braille window's end ends at the end of the buffer. Note that this does mean that _windowRowBufferOffsets now only contains as meny rows as necessary to get to the end of the buffer if the end of the buffer is within the window.

michaelDCurran committed:
* braille: add a filter_displayNumRows extension point, allowing external code to override the number of rows on a braille display, similar to displaySize.
* braille.handler: add a displayNumRows property, similar to the displaySize property, which filters display.numRows via filter_displayNumRows, and if it changes from last time, calls handle_displaySizeChanged extension point, with an extra keyword argument of displayNumRows.
* braille.handler: add a displayNumCols property, which calculates number of columns based on displaySize and displayNumRows.
* braille.BrailleBuffer: calculations that use a display's numrows and numCols now use the handlers's displayNumRows and displayNumCols, so that they can be controled by external code (E.g. remote).
* braille.handler: add a _normalizeCellArraySize method, which takes a list of braille cells of up to length oldCellCount and sequencially formatted into oldNumRows, and reformats the cells list so that it becomes newCellCount in length and sequencially formatted into newNumRows, padding or truncating rows where necessary. Examples:
  - If oldCellCount is 100, oldNumRows is 5 (oldNumCols is 20), and newCellCount is 70, newNumRows is 7 (newNumCols is 10), then the content will have two extra blank rows, and each row will be cut off at 10 cells.
   - If oldCellCount is 70, oldNumRows is 7 (oldNumCols is 10), and newCellCount is 100, newNumRows is 5 (newNumCols is 20), then the last two rows will be cut off, and each row will be padded out by another 10 blank cells.
 * braille.handler._writeCells: replace the old code that padded out the cells to display size, and instead process the cells with _normalizeCellArraySize.

michaelDCurran committed:
Add unit tests for braille.handler._normalizeCellArraySize

michaelDCurran commented:
@LeonarddeR  Thanks for your ideas re displaySize and displayNumRows.
I extended this to also add a filter_displayNumRows extension point similar to displaySize, which means in your example, your machine can in theory report that it has 2 rows of 40 cells. The controlled machine will only see the first row of 40 cells of course.
As we move forward with multiline, I think it is important to maintain row / column layout when remoted, and just pad or truncate rows and columns where necessary, rather than wrapping over rows etc. As eventually we are going to want something like one region per row, and the ability to pan horizontally to see more of all the regions at the same time. 
 A part from other fixes and addition to unit tests, the main changes I made around your ideas were:
* braille: add a filter_displayNumRows extension point, allowing external code to override the number of rows on a braille display, similar to displaySize.
* braille.handler: add a displayNumRows property, similar to the displaySize property, which filters display.numRows via filter_displayNumRows, and if it changes from last time, calls handle_displaySizeChanged extension point, with an extra keyword argument of displayNumRows.
* braille.handler: add a displayNumCols property, which calculates number of columns based on displaySize and displayNumRows.
* braille.BrailleBuffer: calculations that use a display's numrows and numCols now use the handlers's displayNumRows and displayNumCols, so that they can be controled by external code (E.g. remote).
* braille.handler: add a _normalizeCellArraySize method, which takes a list of braille cells of up to length oldCellCount and sequencially formatted into oldNumRows, and reformats the cells list so that it becomes newCellCount in length and sequencially formatted into newNumRows, padding or truncating rows where necessary. Examples:
  - If oldCellCount is 100, oldNumRows is 5 (oldNumCols is 20), and newCellCount is 70, newNumRows is 7 (newNumCols is 10), then the content will have two extra blank rows, and each row will be cut off at 10 cells.
   - If oldCellCount is 70, oldNumRows is 7 (oldNumCols is 10), and newCellCount is 100, newNumRows is 5 (newNumCols is 20), then the last two rows will be cut off, and each row will be padded out by another 10 blank cells.
 * braille.handler._writeCells: replace the old code that padded out the cells to display size, and instead process the cells with _normalizeCellArraySize.

michaelDCurran commented:
As this pr adds a new filter_displayNumRows extension point, it does mean it becomes backward incompatible. However, in the worse case, where the controlled machine had a 10 row display  of 20 cells each, and the controlling machine had a singleline display of 40,  the controling machine will see a single line of 40 made up of 2 rows of 20. 
Of course once the controlling machine was updated and could use the displayNumrows filter, then both machines would correctly see just 1 row of 40 cells.

michaelDCurran committed:
Format lists in unit tests for braille.handler._normalizeCellArraySize  so that they are one row per line, and telling Ruff to not touch them.

LeonarddeR commented on source/braille.py:
```diff
@@ -1760,9 +1760,10 @@ def __init__(self, handler):
-		#: The position in L{brailleCells} where the display window starts (inclusive).
-		#: @type: int
-		self.windowStartPos = 0
+		#: A list representing the rows in the braille window,
+		#: each item being a tuple of start and end braille buffer offsets.
+		#: Splitting the window into independent rows allows for optional avoidance of  splitting words across rows.
+		self._windowRowBufferOffsets: list[tuple[int, int]] = [(0, 0)]
```
Shouldn't this be a variable doc string nowadays?
```suggestion
		self._windowRowBufferOffsets: list[tuple[int, int]] = [(0, 0)]
		"""A list representing the rows in the braille window,
		each item being a tuple of start and end braille buffer offsets.
		Splitting the window into independent rows allows for optional avoidance of  splitting words across rows.
		"""
```

LeonarddeR commented on source/braille.py:
```diff
@@ -2237,6 +2271,14 @@ def formatCellsForLog(cells: List[int]) -> str:
 @type value: int
 """
 
+filter_displayNumRows = extensionPoints.Filter()
+"""
+Filter that allows components or add-ons to change the number of rows used for braille output.
+For example, when a system has a display with 10 rows, but is being controlled by a remote system with a display of 5 rows, the display number of rows should be lowered to 5.
+@param value: the number of rows on the current display.
+@type value: int
+"""
```
```suggestion
filter_displayNumRows = extensionPoints.Filter[int]()
"""
Filter that allows components or add-ons to change the number of rows used for braille output.
For example, when a system has a display with 10 rows, but is being controlled by a remote system with a display of 5 rows, the display number of rows should be lowered to 5.
:param value: the number of rows on the current display.
"""
```

LeonarddeR commented on source/braille.py:
```diff
@@ -2452,6 +2501,42 @@ def _set_displaySize(self, value):
+		Returns the number of columns on the display.
+		This is calculated from displaySize and displayNumRows.
+		"""
+		return self.displaySize // self.displayNumRows
```
Ugh, I'm afraid there is an edge case here. Imagine the case where someone registers a handler for `filter_displayNumRows` without lowering the display size. That could potentially cause the handler to think that there is only one row with 320 cells, for example.
Thinking about that, it should probably be investigated how the braille viewer behaves with all this.

michaelDCurran replied:
What if displayNumRows and displayNumCols began as the display's numRows and numCols respectively, and then were filtered on a new filter_displayDimensions extension point which takes a tuple of numRows and numCols. 
Then displaySize is calculated from displayNumRows and displayNumCols. It is run through the original filter_displaySize extension point. If it is found to change, then we use that instead, and force displayNumRows to 1 and displayNumCols to the filtered displaySize. 
So for older copies of NVDA / remote that only support filter_displaySize, you get one row of displaySize. For modern NVDA / remote you'll get proper rows and columns.
The recommendation would be newer NVDA / remote only registers displayDimentions and not displaySize, though if they did make it be numRows * numCols that would also still be okay.

LeonarddeR replied:
I think this makes sense, though for numRows when filtering with display size, you could also still consider `min(self.display.numRows, math.ceil(self.displaySize / self.display.numCols))`. Or do you think this has issues?

michaelDCurran committed:
Ensure braille messages correctly show on the display:
* BrailleBuffer: abstract  _windowRowBufferOffsets calculation from the windowStartPos setter into its own `_calculateWindowRowBufferOffsets` method, calling it from the windowStartOf setter, but also at the end of BrailleBuffer.update to ensure that the window row buffer offsets are recalculated when the contentchanges.
* BrailleBuffer: always initialize _windowRowBufferOffsets with one row of offsets 0, 1, rather than 0, 0, otherwise windowStartPos could end up being -1.

michaelDCurran committed:
Some rethinking of the braille display size filters:

* braille: Removed filter_displayNumRows.
* braille.handler: removed the displayNumrows and displayNumCols handler properties.
* braille: Added a filter_displayDimensions extension point which takes a namedtuple of numRows and numCols, allowing external code to dictate the number of rows and columns for the display. This should be used now in place of filter_displaySize. This stops the worry of external code changing the number of rows but forgetting to change the over all size.
 * braille.handler: Added a displayDimensions property which:
  - creates a DisplayDimensions namedtuple of the display's real numRows and numCols.
  - Filters this with filter_displayDimensions.
  - Calculates displaySize from this (multiplying the filtered numRows and numCols).
  - Filters this with the old filter_displaySize.
  - If the displaySize changes due to filtering, for backwards compatibility the DisplayDimensions numRows is set to 1 and numCols is set to the filtered displaySize.
  - If the DisplayDimensions have changed at all from the internal cache after all the filters and calculations, then fire the handle_displaySizechanged extension point passing displaySize, numRows and numCols.
* Use displayDimensions.numrows and displayDimensions.numCols in all braille handler code  in place of the removed displayNumRows and displayNumCols properties respectively.

  This all means that new NVDA / remote code can specify full display dimensions, but for backwards compatibility, if the old displaySize filter is used, then braille is forced to a single row with displaySize cells.

michaelDCurran committed:
braille.handler.displayDimensions: only try and apply the displaySize filter and possible force the display to one row if at least one handler has registered for the displaySize filter. this fixes the dge case where old singleline external code happened to want the same displaySize as NVDA's current multiline display (I.e. numRows > 1 and numRows * numCols == external's displaySize).

michaelDCurran committed:
Braille unit tests that check the old internal _displaySize variable should now use _displayDimensions.

pre-commit-ci[bot] committed:
Pre-commit auto-fix

LeonarddeR commented on source/braille.py:
```diff
@@ -2228,13 +2272,24 @@ def formatCellsForLog(cells: List[int]) -> str:
+If this filter is used, NVDA will assume that the display has 1 row of `displaySize` cells.
+"""
+
+DisplayDimensions = NamedTuple("DisplayDimensions", [("numRows", int), ("numCols", int)])
```
Have you considered this syntax, which is a bit cleaner IMO:
```suggestion
class DisplayDimensions(NamedTuple):
	numRows: int
	numCols: int

	@property
	def displaySize(self):
		return self.numCols * self.numRows
```

michaelDCurran replied:
Thanks, I like that. I can Stop writing `numRows * numCols` everywhere.

LeonarddeR commented on source/braille.py:
```diff
@@ -2430,26 +2489,65 @@ def _get_shouldAutoTether(self) -> bool:
+			numCols=self.display.numCols if self.display else 0,
+		)
+		filteredDisplayDimensions = filter_displayDimensions.apply(rawDisplayDimensions)
+		# Would be nice if there were a more official way to find out if the displaySize filter is currently registered by at least 1 handler.
```
It should be pretty trivial to extend extensionPoints.util with an additional property that returns `len(self._handlers)`.

LeonarddeR reviewed (commented):
I like it!

michaelDCurran committed:
braille: improve syntax of DisplayDimensions namedtuple and add displaySize helper property.

michaelDCurran committed:
Braille unit test now uses DisplayDimension.displaySize.

seanbudd commented on source/braille.py:
```diff
@@ -2430,26 +2497,64 @@ def _get_shouldAutoTether(self) -> bool:
+		filteredDisplayDimensions = filter_displayDimensions.apply(rawDisplayDimensions)
+		# Would be nice if there were a more official way to find out if the displaySize filter is currently registered by at least 1 handler.
+		calculatedDisplaySize = filteredDisplayDimensions.displaySize
+		if len(list(filter_displaySize.handlers)):
```
I agree with the comment here as an alternative https://github.com/nvaccess/nvda/pull/17011/files#r1723652836

michaelDCurran replied:
Ah, I did not know that `next` took an extra argument. Thanks!
Re abstracting this, I'd prefer to do it after a bit more consideration and in a separate pr if we do.

seanbudd commented on source/braille.py:
```diff
@@ -1760,9 +1760,10 @@ def __init__(self, handler):
-		self.windowStartPos = 0
+		#: A list representing the rows in the braille window,
+		#: each item being a tuple of start and end braille buffer offsets.
+		#: Splitting the window into independent rows allows for optional avoidance of  splitting words across rows.
```
```suggestion
		#: Splitting the window into independent rows allows for optional avoidance of splitting words across rows.
```

seanbudd commented on tests/unit/test_braille/test_normalizeCellArraySize.py:
```diff
@@ -0,0 +1,143 @@
+		oldCells = [1, 2, 3, 4, 5, 6, 7, 8]  # fmt: skip
+		expectedNewCells = [1, 2, 3, 4]  # fmt: skip
+		assert braille.handler is not None
+		newcells = braille.handler._normalizeCellArraySize(oldCells, 8, 1, 4, 1)
```
could you add param info here? it makes understanding what the tests are doing much easier

```suggestion
		newcells = braille.handler._normalizeCellArraySize(oldCells, oldCellCount=8, oldNumRows=1, newCellCount=4, newNumRows=1)
```

michaelDCurran commented on source/braille.py:
```diff
@@ -2244,6 +2307,10 @@ def formatCellsForLog(cells: List[int]) -> str:
 The local system should be notified about display size changes at the remote system.
 @param displaySize: The current display size used by the braille handler.
 @type displaySize: int
+@param displayDimensions.numRows: The current number of rows used by the braille handler.
+@type displayDimensions.numRows: int
+@param displayDimensions.numCols: The current number of columns used by the braille handler.
+@type displayDimensions.numCols: int
```
Extension points only pass keyword arguments supported by the handler to the hander. See `extensionPoints.utils.callWithSupportedKwargs`. 
Yes, there are new keyword arguments on this action a handler can take advantage of, but existing handlers will not break. No existing keyword arguments were removed.

michaelDCurran committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <sean@nvaccess.org>

michaelDCurran committed:
Merge branch 'master' into brailleWordWrap

michaelDCurran committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <sean@nvaccess.org>

michaelDCurran committed:
Add type info

michaelDCurran committed:
Add keyword arguments to braille test_normalizeCellArraySize.

pre-commit-ci[bot] committed:
Pre-commit auto-fix

michaelDCurran committed:
Developer guide: list filter_displayDimensions

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -1313,7 +1313,8 @@ For examples of how to define and use new extension points, please see the code
 |---|---|---|
-|`Filter` |`filter_displaySize` |Allows components or add-ons to change the display size used for braille output.|
+|`Filter` |`filter_displaySize` | [Deprecated] Allows components or add-ons to change the display size used for braille output.|
+|`Filter` |`filter_displayDimensions` | Allows components or add-ons to change the number of rows and columns  of the display used for braille output.|
```
```suggestion
|`Filter` |`filter_displayDimensions` | Allows components or add-ons to change the number of rows and columns of the display used for braille output.|
```

seanbudd commented on source/braille.py:
```diff
@@ -1857,28 +1858,63 @@ def bufferPositionsToRawText(self, startPos, endPos):
+			return min(start + col, end - 1)
+		raise ValueError("Position outside window")
+
+	def _get_windowStartPos(self):
+		return self.windowPosToBufferPos(0)
+
+	def _set_windowStartPos(self, pos):
+		self._calculateWindowRowBufferOffsets(pos)
```
can these get type hints added

seanbudd committed:
Update projectDocs/dev/developerGuide/developerGuide.md

michaelDCurran committed:
braille.py: add type info.

pre-commit-ci[bot] committed:
Pre-commit auto-fix

seanbudd commented on source/braille.py:
```diff
@@ -1760,9 +1760,10 @@ def __init__(self, handler):
-		#: The position in L{brailleCells} where the display window starts (inclusive).
-		#: @type: int
-		self.windowStartPos = 0
+		#: A list representing the rows in the braille window,
+		#: each item being a tuple of start and end braille buffer offsets.
+		#: Splitting the window into independent rows allows for optional avoidance of splitting words across rows.
+		self._windowRowBufferOffsets: list[tuple[int, int]] = [(0, 1)]
```
this should probably be a doc string
```suggestion
		self._windowRowBufferOffsets: list[tuple[int, int]] = [(0, 1)]
		"""
		A list representing the rows in the braille window,
		each item being a tuple of start and end braille buffer offsets.
		Splitting the window into independent rows allows for optional avoidance of splitting words across rows.
		"""
```

seanbudd commented on source/braille.py:
```diff
@@ -2228,13 +2278,32 @@ def formatCellsForLog(cells: List[int]) -> str:
+"""
+Filter that allows components or add-ons to change the number of rows and columns used for braille output.
+For example, when a system has a display with 10 rows and 20 columns, but is being controlled by a remote system with a display of 5 rows and 40 coluns, the display number of rows should be lowered to 5.
+@param value: a DisplayDimensions namedtuple with the number of rows and columns of the current display.
```
```suggestion
:param value: a DisplayDimensions namedtuple with the number of rows and columns of the current display.
```

seanbudd commented on source/braille.py:
```diff
@@ -2244,6 +2313,10 @@ def formatCellsForLog(cells: List[int]) -> str:
 The local system should be notified about display size changes at the remote system.
 @param displaySize: The current display size used by the braille handler.
 @type displaySize: int
+@param displayDimensions.numRows: The current number of rows used by the braille handler.
+@type displayDimensions.numRows: int
+@param displayDimensions.numCols: The current number of columns used by the braille handler.
+@type displayDimensions.numCols: int
```
```suggestion
:param displaySize: The current display size used by the braille handler.
:type displaySize: int
:param displayDimensions.numRows: The current number of rows used by the braille handler.
:type displayDimensions.numRows: int
:param displayDimensions.numCols: The current number of columns used by the braille handler.
:type displayDimensions.numCols: int
```

seanbudd reviewed (commented):
Apologies for the late extra review - caught these when reviewing unresolved conversations

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -1313,7 +1313,8 @@ For examples of how to define and use new extension points, please see the code
 | Type |Extension Point |Description|
 |---|---|---|
-|`Filter` |`filter_displaySize` |Allows components or add-ons to change the display size used for braille output.|
+|`Filter` |`filter_displaySize` | [Deprecated] Allows components or add-ons to change the display size used for braille output.|
```
can this get a deprecation notice added to the change log?

michaelDCurran committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <sean@nvaccess.org>

michaelDCurran committed:
Merge branch 'master' into brailleWordWrap

michaelDCurran committed:
Update what's new

pre-commit-ci[bot] committed:
Pre-commit auto-fix

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -37,6 +37,9 @@ Please open a GitHub issue if your add-on has an issue with updating to the new
 
 #### Deprecations
 
+* The `braille.filter_displaySize` extension point is deprecated. Please use `braille.filter_displayDimensions` instead. (#17011)
+
+## 2024.4
```
```suggestion
* The `braille.filter_displaySize` extension point is deprecated.
Please use `braille.filter_displayDimensions` instead. (#17011)

```

seanbudd committed:
Update user_docs/en/changes.md
